### ISSUE TITLE
[YSQL] "UPDATE" fails if any "NOT NULL" columns aren't updated

### ISSUE THREAD

**Issue by okready** [yugabyte/yugabyte-db#3174]
Attempting to `UPDATE` any rows in a table with `NOT NULL` columns fails if any of those columns aren't included in the `SET` clause. The following example illustrates this issue:
```sql
CREATE TABLE foo (id SERIAL PRIMARY KEY, x INTEGER NOT NULL, y INTEGER NOT NULL);
INSERT INTO foo (x, y) VALUES (2, 3);
UPDATE foo SET y = 4 WHERE id = 1;
```

This succeeds in PostgreSQL 11, with `y` being modified without modifying `x`, but fails in YugaByteDB 2.0.8.0 with the following error:
```
ERROR:  null value in column "x" violates not-null constraint
DETAIL:  Failing row contains (1, null, 4).
```

If the `NOT NULL` constraint is dropped from the `x` column, the `UPDATE` statement works properly (the `x` column value is left intact).

I wasn't sure if this was due to an architectural limitation of YugaByteDB, but I didn't see any mention of it in the YSQL documentation for `UPDATE` or `CREATE TABLE`, nor did I find any existing issues mentioning it.

**Comment by kmuthukk** [comment]
Thx for reporting this issue @okready. 

Yes - this doesn't look right. We'll take a look at this.

**Comment by kmuthukk** [comment]
@m-iancu : Sounds like this recent https://github.com/yugabyte/yugabyte-db/commit/7e222b42902b5d73067e96ce58a26e679f11a618 might also address this issue?

**Comment by m-iancu** [comment]
@kmuthukk Confirmed locally that 7e222b4 does not fix this (it only addresses the execution part).

I suspect the issue is due to the single-row optimization (https://github.com/yugabyte/yugabyte-db/commit/564ddd1470d6e35afb37543b26a464774cafb6f5).
When that kicks in we skip the select and just update directly. Therefore, we don't have the values for the non-updated column in the slot (and constraint checking interprets that as `NULL`).

For the fix, I think we need to look at the actually _modified_ columns in that case (somewhere in: https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/backend/executor/execMain.c#L1994).

**Comment by kmuthukk** [comment]
<< When that kicks in we skip the select  >>

@m-iancu -- if we skip the SELECT, how do we ensure that the row gets updated only if it exists, and that we don't end up creating a non-existent row as a side-effect of the UPDATE.

**Comment by m-iancu** [comment]
@kmuthukk We check in DocDB (effectively we push down the read similar to an `UPDATE IF EXISTS` in CQL).

**Comment by kmuthukk** [comment]
Aah nice. Thx, @m-iancu

**Comment by OlegLoginov** [comment]
Call stack:
```
ExecConstraints(ResultRelInfo * resultRelInfo, TupleTableSlot * slot, EState * estate) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/execMain.c:1998)
ExecUpdate(ModifyTableState * mtstate, ItemPointer tupleid, HeapTuple oldtuple, TupleTableSlot * slot, TupleTableSlot * planSlot, EPQState * epqstate, EState * estate, _Bool canSetTag) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/nodeModifyTable.c:1133)
ExecModifyTable(PlanState * pstate) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/nodeModifyTable.c:2392)
ExecProcNodeFirst(PlanState * node) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/execProcnode.c:446)
ExecProcNode(PlanState * node) (/home/oleg/code/src/postgres/src/include/executor/executor.h:247)
ExecutePlan(_Bool execute_once, DestReceiver * dest, ScanDirection direction, uint64 numberTuples, _Bool sendTuples, CmdType operation, _Bool use_parallel_mode, PlanState * planstate, EState * estate) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/execMain.c:1724)
standard_ExecutorRun(QueryDesc * queryDesc, ScanDirection direction, uint64 count, _Bool execute_once) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/execMain.c:365)
yb_pg_metrics.so!ybpgm_ExecutorRun(QueryDesc * queryDesc, ScanDirection direction, uint64 count, _Bool execute_once) (/home/oleg/code/yugabyte-db3/src/postgres/contrib/yb_pg_metrics/yb_pg_metrics.c:402)
ExecutorRun(QueryDesc * queryDesc, ScanDirection direction, uint64 count, _Bool execute_once) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/executor/execMain.c:306)
ProcessQuery(PlannedStmt * plan, const char * sourceText, ParamListInfo params, QueryEnvironment * queryEnv, DestReceiver * dest, char * completionTag, _Bool isSingleRowModifyTxn) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/tcop/pquery.c:171)
PortalRunMulti(Portal portal, _Bool isTopLevel, _Bool setHoldSnapshot, DestReceiver * dest, DestReceiver * altdest, char * completionTag) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/tcop/pquery.c:1306)
PortalRun(Portal portal, long count, _Bool isTopLevel, _Bool run_once, DestReceiver * dest, DestReceiver * altdest, char * completionTag) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/tcop/pquery.c:809)
exec_simple_query(const char * query_string) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/tcop/postgres.c:1159)
yb_exec_simple_query_attempting_to_restart_read(const char * query_string, MemoryContext exec_context) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/tcop/postgres.c:4019)
PostgresMain(int argc, char ** argv, const char * dbname, const char * username) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/tcop/postgres.c:4666)
BackendRun(Port * port) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/postmaster/postmaster.c:4398)
BackendStartup(Port * port) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/postmaster/postmaster.c:4064)
ServerLoop() (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/postmaster/postmaster.c:1724)
PostmasterMain(int argc, char ** argv) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/postmaster/postmaster.c:1387)
PostgresServerProcessMain(int argc, char ** argv) (/home/oleg/code/yugabyte-db3/src/postgres/src/backend/main/main.c:234)
```

**Comment by m-iancu** [comment]
Fixed by 07f07aa.
